pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrRuleChainVisitor.java

```diff
@@ -0,0 +1,42 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.antlr;
+
+import java.util.List;
+
+import net.sourceforge.pmd.Rule;
+import net.sourceforge.pmd.RuleContext;
+import net.sourceforge.pmd.lang.ast.AntlrBaseNode;
+import net.sourceforge.pmd.lang.ast.Node;
+import net.sourceforge.pmd.lang.rule.AbstractRuleChainVisitor;
+import net.sourceforge.pmd.lang.rule.XPathRule;
+
+public class AntlrRuleChainVisitor extends AbstractRuleChainVisitor {
+    @Override
+    protected void visit(Rule rule, Node node, RuleContext ctx) {
```

Comment:
Just like I said in my comment above, this class is part of a larger set of duplicates. 

I think it tries to introduce the missing abstraction over AST visitors: it allows writing a single implementation for rulechain visitors. But it's specific to antlr implementations, even though every implementation is coded that way. I think we can go further.